### PR TITLE
Acquire S4U2Self credentials, impersonating a user that is identified by

### PR DESCRIPTION
**Pull request by puran157**
an x509 certificate and/or user name. Implemented in krb5 mechanism.
If only a certificate is passed, client authentication data
is not sent in the request.

Signed-off-by: Isaac Boukris <iboukris@gmail.com>
Signed-off-by: Puran Chand <pchand@vmware.com>

### CODE REVIEW

File: src/lib/gssapi/mechglue/Makefile.in [hunk 1]
```diff
@@ -15,6 +15,7 @@ SRCS = \
 	$(srcdir)/g_acquire_cred.c \
 	$(srcdir)/g_acquire_cred_with_pw.c \
 	$(srcdir)/g_acquire_cred_imp_name.c \
+	$(srcdir)/g_acquire_cred_imp_cert.c \
```
Comment (iboukris) [inline src/lib/gssapi/mechglue/Makefile.in:18]: We need to get rid of these, now that the file is gone.

**Comment by iboukris** [comment]
Now that we use gss_acquire_cred_impersonate_name(), I think we get get rid of t_s4u_cert as well.

Instead add a new `c:` prefix in `tests/gssapi/common.c` so we call something like:
`r.run(['./t_s4u', 'c:' + princ_cert(name), '-', r.keytab])`
Take a look at commit 0b872730081f01d0d2b6f78675bb5c74604cef62 for example.

File: src/tests/gssapi/common.h [hunk 2]
```diff
@@ -51,6 +51,9 @@ void check_k5err(krb5_context context, const char *msg, krb5_error_code code);
 /* Display an error message containing msg and exit. */
 void errout(const char *msg);
 
+/* Import cert data */
+gss_name_t import_cert(gss_buffer_desc buf);
+
```
Comment (iboukris) [inline src/tests/gssapi/common.h:56]: We shouldn't need import_cert(), instead import_name() should interpret the 'c:' prefix.

**Comment by puran157** [comment]
@iboukris I am working on tests, also do I need to handle this new name type in other APIs,
say gss_release_name(), gss_compare_name(), gss_display_name() etc?

**Comment by iboukris** [comment]
Probably yeah, with the current approach we need to at least handle gss_release_name().  But maybe if we store the cert in the krb5 princ that would be simpler, not sure.

**Comment by puran157** [comment]
> Probably yeah, with the current approach we need to at least handle gss_release_name(). But maybe if we store the cert in the krb5 princ that would be simpler, not sure.

Got it, I will see if I can use krb5 pric instead of new data-type.
Can you point me to a documentation for test-setup. I see tests failing for MR but I can't run those locally

**Comment by puran157** [comment]
Also I see tests are failing because they need an actual cert, LMK how I can work on this

**Comment by iboukris** [comment]
Yeah, we don't need import_cert() it is only for kvno, i've forgotten. I'll take a closer look.

**Comment by iboukris** [comment]
@puran157 take a look: https://github.com/iboukris/krb5/commit/634ff72b4cef76a88564df2d6870cdd21be70acd

**Comment by puran157** [comment]
> @puran157 take a look: [iboukris@634ff72](https://github.com/iboukris/krb5/commit/634ff72b4cef76a88564df2d6870cdd21be70acd)

Thanks, merged the changes in this MR

File: src/lib/gssapi/krb5/s4u_gss_glue.c [hunk 3]
```diff
@@ -48,37 +48,44 @@ kg_impersonate_name(OM_uint32 *minor_status,
...
-                                             user->ad_context,
-                                             AD_USAGE_TGS_REQ,
-                                             &in_creds.authdata);
-        if (code != 0) {
-            k5_mutex_unlock(&user->lock);
-            *minor_status = code;
-            return GSS_S_FAILURE;
+    /* Client authroization data can only be sent in the request if a user name
+     * is provided */
+    if (user != NULL) {
```
Comment (iboukris) [inline src/lib/gssapi/krb5/s4u_gss_glue.c:67]: I don't think we need this part, it was a limitation in the old approach.
I think we can reset to my last minimal commit, then think what we can improve, why etc.

Comment (puran157) [reply src/lib/gssapi/krb5/s4u_gss_glue.c]: Fixed

**Comment by iboukris** [comment]
Looks better now, @greghudson any thoughts on this so far?

**Comment by iboukris** [comment]
I think we'd need a new krb5 name type, X500 means "cn=x,dn=y" style, not X509 certificate.

**Comment by puran157** [comment]
> I think we'd need a new krb5 name type, X500 means "cn=x,dn=y" style, not X509 certificate.

Sure, I will work on it.

**Comment by puran157** [comment]
I see the flow should be (and pls correct me if I am not quite following)
1. AS-REQ/AS-REP for service user to get TGT
2. TGS-REQ/TGS-REP using TGT from step1 and User Principal name of delegation-user to get impersonation token
3. TGS-REQ/TGS-REP with impersonation token and service-principal-name to get kerberos token (to be used while authenticating with service)

I see it working with User-Principal-Name but in case of cert, it triggers AS-REQ/AS-REP as step 2 with certificate data.

**Comment by iboukris** [comment]
> I see it working with User-Principal-Name but in case of cert, it triggers AS-REQ/AS-REP as step 2 with certificate data.

Between 1) and 2) there is a stage of realm lookup involving AS-REQ/AS-REP, but it only happens if you use enterprise principal or client-cert to identify the delegated user.

Can you elaborate on your testing? For certificate to work you may need some mapping configuration in the AD side, I don't recall the details so I'd recommend testing with a Windows client ans see that it works before trying the MIT client.

**Comment by puran157** [comment]
> Can you elaborate on your testing? For certificate to work you may need some mapping configuration in the AD side, I don't recall the details so I'd recommend testing with a Windows client ans see that it works before trying the MIT client

I see, its performing a realm lookup since I am passing just the client-certificate, the AD fails with KRB5KDC_ERR_C_PRINCIPAL_UNKNOWN. Seems like a missing mapping on AD side. I will dig more into this

**Comment by puran157** [comment]
> I think we'd need a new krb5 name type, X500 means "cn=x,dn=y" style, not X509 certificate.

Do we have any type as per spec which we can use for certificate. Not sure if spec will have a role to play here, AFAIK, AS-REQ anyway has empty cname with type set to X500 while sending just the client-cert.

See s4u_creds.c::s4u_identify_user()
```
    /*
     * Windows clients send the certificate subject as the client name.
     * However, Windows KDC seem to be happy with an empty string as long as
     * the name-type is NT-X500-PRINCIPAL.
     */
    client.data = &empty_name;
    client.length = 1;
    client.type = KRB5_NT_X500_PRINCIPAL;
```
I believe new name_type will be used internally to make a decision to pass principal or the certificate data from principal structure.

**Comment by iboukris** [comment]
> I see, its performing a realm lookup since I am passing just the client-certificate, the AD fails with KRB5KDC_ERR_C_PRINCIPAL_UNKNOWN. 

Try with 'kvno -F' with a certificate file first, when you are testing.

> AFAIK, AS-REQ anyway has empty cname with type set to X500 while sending just the client-cert.

There the usage of X500 name type is right, because it supposed to be the cert_subject, but to represent an actual cert-der data we may want a new type.

**Comment by puran157** [comment]
> I think we'd need a new krb5 name type, X500 means "cn=x,dn=y" style, not X509 certificate.

Fixed.

**Comment by greghudson** [comment]
I think this just needs minor adjustments, which I can take care of.

Please make sure to call attention in a pull request when you make up an OID value in the code, so it can be properly assigned to avoid conflicts.  This value looks okay, but needs to be recorded to avoid future conflicts, and a comment needs to be added where it is declared.

The name type uses a negative value adjacent to values apparently used by Microsoft.  Although this name type shouldn't appear on the wire, it could easily conflict with another Microsoft extension in the future.  I will probably pick a large negative number, especially as there's no concern over the DER encoding size of the integer.

**Comment by iboukris** [comment]
> The name type uses a negative value adjacent to values apparently used by Microsoft. Although this name type shouldn't appear on the wire, it could easily conflict with another Microsoft extension in the future. I will probably pick a large negative number, especially as there's no concern over the DER encoding size of the integer.

I now think maybe it would be simpler to avoid the new krb5 name type and use a normal princ type instead, but add a boolean flag to krb5_gss_cred_id_t to note that this is cert request.

**Comment by greghudson** [comment]
Updated per Isaac's suggestion.  Also added documentation and made style changes.

**Comment by iboukris** [comment]
Looks good to me, thanks for the rework!

File: src/lib/gssapi/krb5/import_name.c [hunk 4]
```diff
@@ -188,6 +188,14 @@ krb5_gss_import_name(minor_status, input_name_buffer,
                                    &princ);
         if (code)
             goto cleanup;
+    } else if ((input_name_type != NULL) &&
```
Comment (frozencemetery) [inline src/lib/gssapi/krb5/import_name.c:191]: Parens are superflous, if another round of changes happen.

**Pull request merged**